### PR TITLE
Test parameter parsing out of framework

### PR DESCRIPTION
**Pull request by jnm2**
Umbrella: https://github.com/nunit/nunit/issues/1941

### CODE REVIEW

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs [hunk 1]
```diff
@@ -124,20 +124,25 @@ public ITest Build(string assemblyName, IDictionary<string, object> options)
...
         private TestSuite Build(Assembly assembly, string assemblyPath, IDictionary<string, object> options)
         {
-            TestSuite testAssembly = null;
-
             try
             {
-                if (options.ContainsKey(FrameworkPackageSettings.DefaultTestNamePattern))
-                    TestNameGenerator.DefaultTestNamePattern = options[FrameworkPackageSettings.DefaultTestNamePattern] as string;
+                var testNamePattern = GetValueOrDefault(options, FrameworkPackageSettings.DefaultTestNamePattern) as string;
+                if (testNamePattern != null) TestNameGenerator.DefaultTestNamePattern = testNamePattern;
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: Please don't arbitrarily change code that isn't relevant to the thing you are working on based on personal stylistic preferences.

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs [hunk 2]
```diff
@@ -124,20 +124,25 @@ public ITest Build(string assemblyName, IDictionary<string, object> options)
...
-                    string parameters = options[FrameworkPackageSettings.TestParameters] as string;
-                    if (!string.IsNullOrEmpty(parameters))
-                        foreach (string param in parameters.Split(new[] { ';' }))
+                    // Kept for backwards compatibility with old runners
+                    var testParametersString = GetValueOrDefault(options, FrameworkPackageSettings.TestParameters) as string;
+                    if (!string.IsNullOrEmpty(testParametersString))
+                        foreach (var param in testParametersString.Split(';'))
                         {
-                            int eq = param.IndexOf("=");
+                            var eq = param.IndexOf("=");
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: We're generally conservative about var and use int in a case like this. See the coding standards. Are you using some 3rd party software that makes recommendations about things like this?

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs [hunk 3]
```diff
@@ -148,22 +153,24 @@ private TestSuite Build(Assembly assembly, string assemblyPath, IDictionary<stri
                             }
                         }
                 }
+                
+                var fixtures = GetFixtures(assembly, GetValueOrDefault(options, FrameworkPackageSettings.LOAD) as IList);
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: Again... what does this have to do with the issue?

File: src/NUnitFramework/nunitlite/TextRunner.cs [hunk 4]
```diff
@@ -361,12 +363,52 @@ private int ExploreTests()
                 runSettings[FrameworkPackageSettings.DefaultTestNamePattern] = options.DefaultTestNamePattern;
 
             if (options.TestParameters != null)
-                runSettings[FrameworkPackageSettings.TestParameters] = options.TestParameters;
+            {
+                var dictionary = new Dictionary<string, string>();
```
Comment (CharliePoole) [inline src/NUnitFramework/nunitlite/TextRunner.cs]: As with the console, I'd prefer to see the dictionary creation centralized in the Options hierarchy so that the framework assembly simply receives a dictionary. Otherwise, there is no point in making it a framework run setting. The framework isn't intended to change the settings that are passed to it, just to use them.

As an aside, I suppose that working on so many versions of the same thing is giving both you and the reviewers extra work. 😢

**Comment by CharliePoole** [comment]
An added "learning" here, if you don't mind. Let's say you do a PR that contains code that functionally does three different things. It will only be merged if two committers agree with what you have done for __all three__ things. So as a matter of self-interest, it's best to keep issues focused.

In fact, that was part of your logic (which convinced me!)  in separating what seemed like one issue at first into two. :smile:

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs (hunk 2 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: Yes, ReSharper. The original code mixed var and non-var styles so I figured I'd pick one. Easy enough to undo.

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs (hunk 1 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: :+1:

**Comment by jnm2** [comment]
This is, I suppose, changing two things: the framework, and nunit. Should I split into two PRs then?

**Comment by jnm2** [comment]
This is, I suppose, changing two things: the framework, and nunit. They are split into two commits. Should I split into two PRs then? I saw all these PRs as a single unit of work, only split because repos. But they are intrinsically separable so perhaps two PRs is better.

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs (hunk 3 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: Consistency, instead of having multiple ways to do the same thing. No problem to change back.

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs (hunk 2 above)
Comment (oznetmaster) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: It is easy to set ReSharper to only suggest/change to var when the assignment meets the criteria of NUnit (the RHS explicitely defines the type of the variable).

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs (hunk 3 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: Consistency, instead of doing the same thing multiple ways. No problem to change back.

Perhaps also I should have copied the original `ContainsKey`...indexer pattern. Writing that code is harder for me because it's not only less efficient, it's less DRY, but I can see how maybe you'd want to discuss that first.

File: src/NUnitFramework/nunitlite/TextRunner.cs (hunk 4 above)
Comment (jnm2) [inline src/NUnitFramework/nunitlite/TextRunner.cs]: I'll simply wait for the conclusion of [these comments](https://github.com/nunit/nunit-console/pull/155#issuecomment-269808071) and pick up here. =)

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs (hunk 2 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: @CharliePoole my bad on missing https://github.com/nunit/docs/wiki/Coding-Standards#use-of-the-var-keyword. Will be more attentive.

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs [hunk 5]
```diff
@@ -79,7 +79,7 @@ public ITest Build(Assembly assembly, IDictionary<string, object> options)
 #else
             log.Debug("Loading {0} in AppDomain {1}", assembly.FullName, AppDomain.CurrentDomain.FriendlyName);
 #endif
-            
+
```
Comment (jnm2) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: @CharliePoole I am not seeing these whitespace changes as they happen. Do you care about these? I can work on it.

File: src/NUnitFramework/nunitlite/TextRunner.cs [hunk 6]
```diff
@@ -361,12 +363,52 @@ private int ExploreTests()
...
+            runSettings[FrameworkPackageSettings.TestParametersDictionary] = testParameters;
+
+            if (testParameters.Count != 0)
+            {
+                // Kept for backwards compatibility with old frameworks.
+                // Reserializes the way old frameworks understand, even if the parsing above is changed.
+                // This reserialization cannot be changed without breaking compatibility with old frameworks.
+
+                var oldFrameworkSerializedParameters = new StringBuilder();
+                foreach (var parameter in testParameters)
```
Comment (jnm2) [inline src/NUnitFramework/nunitlite/TextRunner.cs]: @CharliePoole I'm willing to follow https://github.com/nunit/docs/wiki/Coding-Standards#use-of-the-var-keyword and use `foreach (KeyValuePair<string, string> parameter in testParameters)` here too. Would if be good for me to update this?

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs (hunk 5 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: No, it happens to everyone and if there are too many we just use look at the diff ignoring whitespace.

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs (hunk 2 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: Be sure to pay attention to the most important part:
  https://github.com/nunit/docs/wiki/Coding-Standards#general---please-read-this
especially paragraph 2. 😄

**Comment by CharliePoole** [comment]
This is, I suppose, changing two things: the framework, and nunit. They are split into two commits. Should I split into two PRs then? I saw all these PRs as a single unit of work, only split because repos. But they are intrinsically separable so perhaps two PRs is better

No, the "thing" that's being changed is "How nunitlite passes the run parameters to the framework" It requires a change in two places, but it's one issue in one repository.

OTOH, changing the same thing for the console crosses an arbitrary administrative boundary we have created... between the team of people who work on the framework and the team that works on the engine and console. So it needs to be two issues. If too many things cross that boundary and cause us problems, it's in the realm of the possible that we could re-examine the decision. But for now, it sticks.

For a newcomer, this is probably hard to see because we created the boundary but for the moment the same people are working on both sides of it. We hope that won't last and that more new people will come to each project, allowing them to move ahead independently. IOW it's a bet we are making.

File: src/NUnitFramework/nunitlite/TextRunner.cs (hunk 6 above)
Comment (CharliePoole) [inline src/NUnitFramework/nunitlite/TextRunner.cs]: The standard is deliberately vague. IOW, we could have written "only use var if the rhs contains the name of the type" but we didn't. The idea (my idea) is that developers are supposed to use judgement - a notion that is too often forgotten. Whether it's sufficiently clear is a team judgement, since all code is clear to the guy who wrote it (at least for a while). My judgement is that it makes sense in context. Somebody else may say different.

File: src/NUnitFramework/nunitlite/TextRunner.cs (hunk 4 above)
Comment (CharliePoole) [inline src/NUnitFramework/nunitlite/TextRunner.cs]: Fair enough.

**Comment by jnm2** [comment]
> Since it's a common option, it should go into the base class, CommandLineOptions in both cases. If you do it in the framework project first, we may be able to get it into the release this week. The console release will be next week.

Framework done. @CharliePoole The CommandLineOptions route does feel cleaner.

**Comment by jnm2** [comment]
Not quite done, forgot to check in the separate commit for the parameters test fixes.

**Comment by jnm2** [comment]
Decided it was simpler to put the updated tests in the same NUnitLite commit.

**Comment by rprouse** [comment]
Let me take another look at this in the morning and we can figure out what to do for the release. I plan on branching for the release tomorrow, but I can always rebase last minute to pull this in if it is ready.

**Comment by jnm2** [comment]
@CharliePoole not a problem. I'd like to clean that up before merge.

**Comment by jnm2** [comment]
@CharliePoole @rprouse Should be ready.

**Comment by CharliePoole** [comment]
This looks good to me but it's sufficiently important, affecting console, gui, adapter, etc., that I'd like to get another set of eyes on it as well. @ChrisMaddock can you take a look? If we can get this into the release, then the corresponding runner changes will also be able to go in.

File: src/NUnitFramework/framework/Api/FrameworkPackageSettings.cs [hunk 7]
```diff
@@ -106,10 +106,15 @@ public static class FrameworkPackageSettings
         public const string DefaultTestNamePattern = "DefaultTestNamePattern";
 
         /// <summary>
-        /// Parameters to be passed on to the test
+        /// Parameters to be passed on to the tests, serialized to a single string which needs parsing. Obsoleted by <see cref="TestParametersDictionary"/>; kept for backward compatibility.
         /// </summary>
         public const string TestParameters = "TestParameters";
 
+        /// <summary>
+        /// Parameters to be passed on to the tests, already parsed into an IDictionary&lt;string, string>. Replaces <see cref="TestParameters"/>.
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Api/FrameworkPackageSettings.cs]: Just a little typo here. Not important enough to hold up the release, but perhaps can be fixed if you get back to it before it's merged, @jnm2 . 😄

File: src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs [hunk 8]
```diff
@@ -131,26 +131,41 @@ private TestSuite Build(Assembly assembly, string assemblyPath, IDictionary<stri
...
-                            int eq = param.IndexOf("=");
+                    var testParametersDictionary = options[FrameworkPackageSettings.TestParametersDictionary] as IDictionary<string, string>;
+                    if (testParametersDictionary != null)
+                    {
+                        foreach (var parameter in testParametersDictionary)
+                            TestContext.Parameters.Add(parameter.Key, parameter.Value);
+                    }
+                }
+                else
+                {
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs]: Just a suggestion - would this be clearer as an `if...else if` now the noise has been removed? I think that's me going well into the grounds of stylistic preference by now though - so feel free to ignore that. 😄

**Comment by CharliePoole** [comment]
I see that the release branch has already been created so I'm merging this to master. It will be up to @rprouse to decide whether to update th branch with all our merged fixes or to cherry pick them.

File: src/NUnitFramework/framework/Api/FrameworkPackageSettings.cs (hunk 7 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Api/FrameworkPackageSettings.cs]: Ugh! Where? I'm not seeing it!

Comment (CharliePoole) [inline src/NUnitFramework/framework/Api/FrameworkPackageSettings.cs]: It's merged in any case.

Comment (jnm2) [inline src/NUnitFramework/framework/Api/FrameworkPackageSettings.cs]: I know but it'll drive me crazy. How am I not seeing the typo?

Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Api/FrameworkPackageSettings.cs]: Sorry - it's not! I presumed `&lt;string, string>` should be either all escaped or not escaped, but I just tried my 'correction' - and it doesn't compile, as I presume you already knew. Sorry, false alarm!

Comment (jnm2) [inline src/NUnitFramework/framework/Api/FrameworkPackageSettings.cs]: Oh no problem! I would have thought `&gt;` would compile. Weird.

Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Api/FrameworkPackageSettings.cs]: Probably - I only tried `<string, string>` which didn't. :smile:

Comment (jnm2) [inline src/NUnitFramework/framework/Api/FrameworkPackageSettings.cs]: I would have done the whole `<see cref="IDictionary{X,Y}"/>` but there is no facility for specifying closed generics, so users would have seen `IDictionary<TKey, TValue>` no matter what I wrote for `X` and `Y`. :(
